<center><h1>Digits classification SVM</h1></center>

In [1]:
import numpy as np
import pandas as pd
from sklearn.svm import SVC
from sklearn.preprocessing import StandardScaler
from sklearn import metrics
from sklearn.model_selection import KFold,GridSearchCV,train_test_split
import matplotlib.pyplot as plt
import seaborn as sns


In [2]:
digits_train=pd.read_csv("train.csv")

In [4]:
digits_train.head()

label  pixel0  pixel1  pixel2  pixel3  pixel4  pixel5  pixel6  pixel7  \
0      1       0       0       0       0       0       0       0       0   
1      0       0       0       0       0       0       0       0       0   
2      1       0       0       0       0       0       0       0       0   
3      4       0       0       0       0       0       0       0       0   
4      0       0       0       0       0       0       0       0       0   

   pixel8  ...  pixel774  pixel775  pixel776  pixel777  pixel778  pixel779  \
0       0  ...         0         0         0         0         0         0   
1       0  ...         0         0         0         0         0         0   
2       0  ...         0         0         0         0         0         0   
3       0  ...         0         0         0         0         0         0   
4       0  ...         0         0         0         0         0         0   

   pixel780  pixel781  pixel782  pixel783  
0         0         0         0         0  
1         0         0         0         0  
2         0         0         0         0  
3         0         0         0         0  
4         0         0         0         0  

[5 rows x 785 columns]

<center><h1>EDA</h1></center>

In [5]:
digits_train.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 42000 entries, 0 to 41999
Columns: 785 entries, label to pixel783
dtypes: int64(785)
memory usage: 251.5 MB


In [6]:
digits_train.describe()

label   pixel0   pixel1   pixel2   pixel3   pixel4   pixel5  \
count  42000.000000  42000.0  42000.0  42000.0  42000.0  42000.0  42000.0   
mean       4.456643      0.0      0.0      0.0      0.0      0.0      0.0   
std        2.887730      0.0      0.0      0.0      0.0      0.0      0.0   
min        0.000000      0.0      0.0      0.0      0.0      0.0      0.0   
25%        2.000000      0.0      0.0      0.0      0.0      0.0      0.0   
50%        4.000000      0.0      0.0      0.0      0.0      0.0      0.0   
75%        7.000000      0.0      0.0      0.0      0.0      0.0      0.0   
max        9.000000      0.0      0.0      0.0      0.0      0.0      0.0   

        pixel6   pixel7   pixel8  ...      pixel774      pixel775  \
count  42000.0  42000.0  42000.0  ...  42000.000000  42000.000000   
mean       0.0      0.0      0.0  ...      0.219286      0.117095   
std        0.0      0.0      0.0  ...      6.312890      4.633819   
min        0.0      0.0      0.0  ...      0.000000      0.000000   
25%        0.0      0.0      0.0  ...      0.000000      0.000000   
50%        0.0      0.0      0.0  ...      0.000000      0.000000   
75%        0.0      0.0      0.0  ...      0.000000      0.000000   
max        0.0      0.0      0.0  ...    254.000000    254.000000   

           pixel776     pixel777      pixel778      pixel779  pixel780  \
count  42000.000000  42000.00000  42000.000000  42000.000000   42000.0   
mean       0.059024      0.02019      0.017238      0.002857       0.0   
std        3.274488      1.75987      1.894498      0.414264       0.0   
min        0.000000      0.00000      0.000000      0.000000       0.0   
25%        0.000000      0.00000      0.000000      0.000000       0.0   
50%        0.000000      0.00000      0.000000      0.000000       0.0   
75%        0.000000      0.00000      0.000000      0.000000       0.0   
max      253.000000    253.00000    254.000000     62.000000       0.0   

       pixel781  pixel782  pixel783  
count   42000.0   42000.0   42000.0  
mean        0.0       0.0       0.0  
std         0.0       0.0       0.0  
min         0.0       0.0       0.0  
25%         0.0       0.0       0.0  
50%         0.0       0.0       0.0  
75%         0.0       0.0       0.0  
max         0.0       0.0       0.0  

[8 rows x 785 columns]

In [7]:
digits_train.head()

label  pixel0  pixel1  pixel2  pixel3  pixel4  pixel5  pixel6  pixel7  \
0      1       0       0       0       0       0       0       0       0   
1      0       0       0       0       0       0       0       0       0   
2      1       0       0       0       0       0       0       0       0   
3      4       0       0       0       0       0       0       0       0   
4      0       0       0       0       0       0       0       0       0   

   pixel8  ...  pixel774  pixel775  pixel776  pixel777  pixel778  pixel779  \
0       0  ...         0         0         0         0         0         0   
1       0  ...         0         0         0         0         0         0   
2       0  ...         0         0         0         0         0         0   
3       0  ...         0         0         0         0         0         0   
4       0  ...         0         0         0         0         0         0   

   pixel780  pixel781  pixel782  pixel783  
0         0         0         0         0  
1         0         0         0         0  
2         0         0         0         0  
3         0         0         0         0  
4         0         0         0         0  

[5 rows x 785 columns]

<center><h1>Data preparation</h1></center>

In [8]:
Y=digits_train.pop("label")

In [12]:
scaller =StandardScaler()

In [13]:
X=scaller.fit_transform(digits_train)

In [14]:
Xtrain,Xtest,Ytrain,Ytest=train_test_split(X,Y,test_size=0.8,random_state=100)

<center><h1>Model Building</h1></center>

In [15]:
#checking whether a linear model will be better against non linear or not 
model=SVC(kernel='linear')

In [16]:
model.fit(Xtrain,Ytrain)

SVC(kernel='linear')

In [17]:
Y_pred=model.predict(Xtest)

In [18]:
metrics.accuracy_score(y_true=Ytest,y_pred=Y_pred)

0.9106845238095238

In [19]:
model=SVC(kernel='rbf')

In [20]:
model.fit(Xtrain,Ytrain)

SVC()

In [21]:
Y_pred=model.predict(Xtest)

In [22]:
metrics.accuracy_score(y_true=Ytest,y_pred=Y_pred)

0.9337797619047619

Observation:- Turns out the non linear model performs better in terms of accuracy of the model ,Hence we will go with the non linear radial basis kernel

<center><h1>Hyperparameter tuning</h1></center>

In [23]:
folds=KFold(n_splits=5,shuffle=True,random_state=4)

In [24]:
hyperparameter={'gamma':[1e-1,1e-2,1e-3,1e-4],
                'C':[0.1,1,10,100,1000]}

In [25]:
model=SVC(kernel='rbf')

In [27]:
Gridmodel=GridSearchCV(estimator=model,return_train_score=True,verbose=1,scoring='accuracy',cv=folds,param_grid=hyperparameter)

In [28]:
Gridmodel.fit(Xtrain,Ytrain)

Fitting 5 folds for each of 20 candidates, totalling 100 fits


GridSearchCV(cv=KFold(n_splits=5, random_state=4, shuffle=True),
             estimator=SVC(),
             param_grid={'C': [0.1, 1, 10, 100, 1000],
                         'gamma': [0.1, 0.01, 0.001, 0.0001]},
             return_train_score=True, scoring='accuracy', verbose=1)

In [29]:
Gridmodel.best_params_

{'C': 10, 'gamma': 0.001}

Observation-: we got C=10 and 0.001 as the best parameter , thus going forward we will use those values to get a optimal classifier .

<center><h1>Building final model with optimal hyperparameters</h1></center>

In [30]:
model=SVC(kernel='rbf',C=10,gamma=0.001)

In [31]:
model.fit(Xtrain,Ytrain)

SVC(C=10, gamma=0.001)

In [32]:
Y_pred=model.predict(Xtest)

In [33]:
metrics.accuracy_score(y_true=Ytest,y_pred=Y_pred)

0.9433928571428571

<center><h1>Fitting final model on the test data and finding the prediction</h1></center>

In [82]:
digits_test=pd.read_csv("test.csv")

In [83]:
digits_test

pixel0  pixel1  pixel2  pixel3  pixel4  pixel5  pixel6  pixel7  pixel8  \
0           0       0       0       0       0       0       0       0       0   
1           0       0       0       0       0       0       0       0       0   
2           0       0       0       0       0       0       0       0       0   
3           0       0       0       0       0       0       0       0       0   
4           0       0       0       0       0       0       0       0       0   
...       ...     ...     ...     ...     ...     ...     ...     ...     ...   
27995       0       0       0       0       0       0       0       0       0   
27996       0       0       0       0       0       0       0       0       0   
27997       0       0       0       0       0       0       0       0       0   
27998       0       0       0       0       0       0       0       0       0   
27999       0       0       0       0       0       0       0       0       0   

       pixel9  ...  pixel774  pixel775  pixel776  pixel777  pixel778  \
0           0  ...         0         0         0         0         0   
1           0  ...         0         0         0         0         0   
2           0  ...         0         0         0         0         0   
3           0  ...         0         0         0         0         0   
4           0  ...         0         0         0         0         0   
...       ...  ...       ...       ...       ...       ...       ...   
27995       0  ...         0         0         0         0         0   
27996       0  ...         0         0         0         0         0   
27997       0  ...         0         0         0         0         0   
27998       0  ...         0         0         0         0         0   
27999       0  ...         0         0         0         0         0   

       pixel779  pixel780  pixel781  pixel782  pixel783  
0             0         0         0         0         0  
1             0         0         0         0         0  
2             0         0         0         0         0  
3             0         0         0         0         0  
4             0         0         0         0         0  
...         ...       ...       ...       ...       ...  
27995         0         0         0         0         0  
27996         0         0         0         0         0  
27997         0         0         0         0         0  
27998         0         0         0         0         0  
27999         0         0         0         0         0  

[28000 rows x 784 columns]

In [84]:
digits_test=scaller.transform(digits_test)

In [85]:
len(digits_test)

28000

In [86]:
Y_test_pred=model.predict(digits_test)

In [87]:
len(Y_test_pred)

28000

In [88]:
df=pd.DataFrame(columns=['ImageId','Label'])

In [89]:
df['ImageId']=np.arange(1,28001)
df['Label']=Y_test_pred

In [90]:
df=df.set_index('ImageId')

In [91]:
df

Label
ImageId       
1            2
2            0
3            9
4            4
5            3
...        ...
27996        9
27997        7
27998        3
27999        9
28000        2

[28000 rows x 1 columns]

In [93]:
df.to_csv('Submission_1.csv')

Conclusion-:Got 94% accuracy on the test data set on kaggle.